service/interpreter/stateExecutionCounter.go

```diff
@@ -114,10 +168,7 @@ func (e *StateExecutionCounter) updateStateIdSearchAttribute() error {
 	for sid := range e.stateIdCurrentlyExecutingCounts {
 		executingStateIds = append(executingStateIds, sid)
 	}
-	config := e.configer.Get()
-	if config.GetDisableSystemSearchAttribute() {
-		return nil
-	}
+
 	if e.globalVersioner.IsAfterVersionOfOptimizedUpsertSearchAttribute() && len(executingStateIds) == 0 {
```

Comment:
This optimization is broken already for current default/ENABLED_FOR_ALL, and it won't work for ENABLED_FOR_STATES_WITH_WAIT_UNTIL. We need to fix it, and have to update `ClearExecutingStateIdsSearchAttributeFinally` as well.

To keep determinsm, we have to add a `IsBeforeVersionOfExecutingStateIdMode` in globalVersioner. 
Then here:
```
if e.globalVersioner.IsAfterVersionOfOptimizedUpsertSearchAttribute() && 
e.globalVersioner.IsBeforeVersionOfExecutingStateIdMode() &&
len(executingStateIds) == 0 {
    // explain why we stopped using this optimization
    return nil
}
```

Similarly in ClearExecutingStateIdsSearchAttributeFinally:
```
if e.globalVersioner.IsAfterVersionOfOptimizedUpsertSearchAttribute() && 
e.globalVersioner.IsBeforeVersionOfExecutingStateIdMode() {
   if if config.GetDisableSystemSearchAttribute() {	
		return
   }
   upsertSearchAttribute( []string{}) // clear the value.
}
```

Btw, you can also create a method in globalVersioner `IsAfterVersionOfBrokenOptimizedUpsertSearchAttributeAndBeforeVersionOfExecutingStateIdMode()` to combinet this check. Maybe that will look better.

Reply:
Btw, I realized that the correct way to optimize this UpsertSearchAttribute(which is a temporal action) is to check the next state like this: https://github.com/indeedeng/iwf/issues/454

We can also do it in next PR -- as long as we don't release the this PR to production, we don't have to create a new version

Reply:
`IsBeforeVersionOfExecutingStateIdMode` is equal to `!IsAfterVersionOfExecutingStateIdMode`. Do we need this extra method?

The latter method is somehow malformed in your snippet. I changed it to:
```
func (e *StateExecutionCounter) ClearExecutingStateIdsSearchAttributeFinally() {
	config := e.configer.Get()

	if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() {
		if config.GetExecutingStateIdMode() == "DISABLED" {
			return
		}
	} else {
		if config.GetDisableSystemSearchAttribute() {
			return
		}
	}

	if e.globalVersioner.IsAfterVersionOfOptimizedUpsertSearchAttribute() && !e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() && e.totalCurrentlyExecutingCount == 0 {
		err := e.provider.UpsertSearchAttributes(e.ctx, map[string]interface{}{
			service.SearchAttributeExecutingStateIds: []string{},
		})
		if err != nil {
			e.provider.GetLogger(e.ctx).Error("error for upsert SearchAttributeExecutingStateIds", err)
		}
	}
}
```

I believe this ^^ is correct.

And I'd rather work on #454 in a separate branch. I can rebase off of this one and merge #454 into here before merging to main if we want to release other changes.

Reply:
> IsBeforeVersionOfExecutingStateIdMode is equal to !IsAfterVersionOfExecutingStateIdMode. Do we need an extra method for this check?

You are right, we don't have to. Sometimes I just prefer to not use negative statement/expression (reasons are like https://www.obsidiansecurity.com/blog/why-to-stop-writing-negative-code/  ) 

But no much difference and it's not always the case.

Reply:
>And I'd rather work on https://github.com/indeedeng/iwf/issues/454 in a separate branch. I can rebase off of this one and merge https://github.com/indeedeng/iwf/issues/454 into here before merging to main if we want to release other changes. 

+1 
Sounds good. btw maybe not add replay tests for this PR . Otherwise you will have to do it again if we want to reuse the same version (because that optimization is a non-determinstic change)